ucp/useraccounts_module.php

```diff
@@ -319,143 +269,146 @@ function main($id, $mode)
 					}
 
-					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
-
 					// Delete all user badges that are NOT IN the array of marked badges or all if none is selected.
+					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
 					$sql = 'DELETE FROM ' . $user_badge_table . '
 						WHERE user_id = ' . $this_user_id . "
 						$delete_sql";
 					$this->db->sql_query($sql);
 
-					foreach ($markedbadges as $markedbadge)
+					if ($markedbadges)
 					{
-						// Reject duplicates.
-						$sql = "SELECT COUNT(*) as duplcount
-							FROM $user_badge_table
-							WHERE badge_id = $markedbadge
-							AND user_id = $this_user_id";
-						if (!($result = $this->db->sql_query($sql)))
+						// Query valid badges.
+						$sql_array = array(
+							'SELECT' => 'b.badge_id',
+
+							'FROM' => array(
+								$badge_table => 'b'
+							),
+
+							'LEFT_JOIN' => array(
+								array(
+									'FROM' => array($user_badge_table => 'bu'),
+									'ON' => 'b.badge_id = bu.badge_id AND bu.user_id = ' . $this_user_id
+								),
+								array(
+									'FROM' => array($badge_avail_table => 'ba'),
+									'ON' => 'b.badge_id = ba.badge_id AND ba.user_id = ' . $this_user_id
+								),
+							),
+
+							// Badges are valid if they are not already selected (bu.badge_id is null) and are either:
+							//  - a default badge (b.badge_default is true)
+							//  - have been awarded to the user (ba.user_id is not null)
+							'WHERE' => 'bu.badge_id IS NULL AND (b.badge_default = TRUE OR ba.user_id = ' . $this_user_id . ')',
+						);
+
+						$sql = $this->db->sql_build_query('SELECT', $sql_array);
+						$result = $this->db->sql_query($sql);
+
+						// Prepare an array to store all valid badges.
+						$valid_badges = array();
+						while ($row = $this->db->sql_fetchrow($result))
 						{
-							trigger_error($this->user->lang('UCP_DUPLICATE_USER_BADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+							$valid_badges[] = (int)$row['badge_id'];
 						}
-						$duplicates = (int) $this->db->sql_fetchfield('duplcount');
-						$this->db->sql_freeresult($result);
-						if (!$duplicates)
+
+						// Prepare an array for inserting badges.
+						$sql_array = array();
+
+						foreach ($markedbadges as $markedbadge)
 						{
-							// Add a new user badge.
-							$data = array(
-								'user_id' => $this_user_id,
-								'badge_id' => $markedbadge,
-								'badge_order' => null // Order value will be validated after loop.
-							);
-							$sql = 'INSERT INTO ' . $user_badge_table . $this->db->sql_build_array('INSERT', $data);
-							$this->db->sql_query($sql);
+							// Validate requested badges.
+							settype($markedbadge, 'int');
```

Comment:
I expect this is redundant, but probably harmless.

Reply:
My idea here was that we only casted the input from `markedkeys` to an array but not its values to integers, and `$markedbadge` is used in `if (in_array($markedbadge, $valid_badges)).`

Reply:
Two points here:
* `in_array` will only return true if `$markedbadge` is in `$valid_badges`, which we know by definition only contains integers
* `$markedbadges` is used without explicit type checks earlier in the file (but note that `sql_in_set` escapes the entries, so this is safe)